### PR TITLE
Migrate `tanh` from TH to ATen (CUDA)

### LINKED ISSUE #24642: Migrate `tanh` and `tanh_` from the TH to Aten (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### PR DESCRIPTION
**Pull request by kshitij12345**
Closes #24642

Benchmark with same build settings on same system.
gcc : version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)
CUDA : 10.1
GPU : 1050ti

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.tanh(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.tanh(a); torch.cuda.synchronize()',
                              setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")',
                              number=t))
```

Before:

```
torch.tanh(a) a.numel() == 10000 for 20000 times torch.half
0.2816318240002147
torch.tanh(a) a.numel() == 10000 for 20000 times torch.float
0.2728829070001666
torch.tanh(a) a.numel() == 10000 for 20000 times torch.double
0.39797203200214426
torch.tanh(a) a.numel() == 100000 for 20000 times torch.half
0.3228214350019698
torch.tanh(a) a.numel() == 100000 for 20000 times torch.float
0.31780802399953245
torch.tanh(a) a.numel() == 100000 for 20000 times torch.double
1.3745740449994628
```

After:

```
torch.tanh(a) a.numel() == 10000 for 20000 times torch.half
0.27825374500025646
torch.tanh(a) a.numel() == 10000 for 20000 times torch.float
0.27764024499992956
torch.tanh(a) a.numel() == 10000 for 20000 times torch.double
0.3771585260001302
torch.tanh(a) a.numel() == 100000 for 20000 times torch.half
0.2995866400015075
torch.tanh(a) a.numel() == 100000 for 20000 times torch.float
0.28355561699936516
torch.tanh(a) a.numel() == 100000 for 20000 times torch.double
1.393811182002537
```

### CODE REVIEW

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu [hunk 1]
```diff
@@ -92,9 +92,30 @@ void sinh_kernel_cuda(TensorIterator& iter) {
   });
 }
 
+// We manually overload sinh because std::sinh does not work with thrust::complex types.
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:95]: copy-pasted comment, please change sinh to tanh

**Comment by ngimel** [comment]
Since you are adding complex support, please add tests for it in test_complex.py

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu (hunk 1 above)
Comment (kshitij12345) [reply aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:95]: Oops. Will do that. Thanks.

**Comment by kshitij12345** [comment]
@ngimel Thanks for the quick review.

PR doesn't add complex support as the scope is only migrating. So haven't added the complex dispatch and thus no test for complex.

The cuda wrapper specialization for complex is only to be consistent with the current implementations that are already present.

**Comment by ngimel** [comment]
You did add complex dispatch https://github.com/pytorch/pytorch/pull/36995/files#diff-9d11f7cdd593ca7b452bf8e78802afe8R107, so either please remove it (along with thrust::complex workarounds) or test it.

**Review by ngimel — APPROVED** [review]

**Comment by kshitij12345** [comment]
@ngimel Thank You. Copy-Paste issue again.

**Comment by kshitij12345** [comment]
@ngimel Have rebased the branch with latest master :).

**Pull request closed without merging**